internetarchive/utils.py

```diff
@@ -43,7 +45,7 @@
     JSONDecodeError = json.JSONDecodeError  # type: ignore
 
 
-def deep_update(d, u):
+def deep_update(d: dict, u: Mapping) -> dict:
```

Comment:
The recommended way seems to be to use the more abstract type for the arguments and the narrower types for the return values. So `-> dict` would be as recommended but if we follow the Python documentation's recommendation, then `d` should also have the type `Mapping`.

The same is true for all other functions in this file.

Reply:
That sounds very odd to me. The return type here is always `type(d)` since `d` is simply modified in-place and returned...

We could just get rid of the `return` entirely and document it to be an in-place update. The function is only used in one place (`internetarchive.config.get_config`), and the return value is discarded there.

Reply:
I was merely quoting the [relevant part of the documentation](https://docs.python.org/3/library/typing.html#typing.Dict) and it makes a lot of sense to me combined with the duck typing approach of Python, since otherwise you would lose every advantage of duck typing.

With this particular function we have a very special case due to your described in-place mutation, where it would also make sense to keep `dict` as type. Probably it would be best to document the in-place mutation, return nothing, type the function as `-> None` and to use `Mapping` as input type, since that will ensure the appropriate interface for the mutation and the caller will have the exact type information.

Reply:
Modifying a function the returns a value to be a function that returns no value is a breaking change (semver major) so I would prefer to avoid that.

Reply:
It seems very unlikely that someone external is using this function. But you are correct.

Therefore keeping the signature which you already used would be the best course for now.

Reply:
Regarding the more general issue of modifying interfaces, I'm not sure the `utils` module is even considered part of the public API. It isn't mentioned in the docs, for example. Similarly for the CLI, which I can't imagine being intended to be imported from elsewhere. It may be worth moving them to `internetarchive._utils` and `._cli` instead if they should be treated as internal API. (And yeah, such a change should probably happen on a major version increment.)

Reply:
The recommendation in gradual typing is to start by typing the innermost code first and end its the outermost code last.  I do not know this codebase well enough to know which is which but typing is for both internal and external communications.  I started with a file that I was having trouble with on another project.  Next is this file which is relied on by multiple internal files.  What should come next…